runtime/parser2/lexer/lexer.go

```diff
@@ -74,9 +76,47 @@ type lexer struct {
 	startPos position
 }
 
-func Lex(ctx context.Context, input string) chan Token {
+type TokenStream interface {
+	// Next consumes and returns one Token. If there are no tokens remaining, it returns Token{TokenEOF}
+	Next() Token
+	// Close provides an opportunity for a TokenStream implementation to clean-up
+	Close()
+	// Code returns the entire TokenStream in source code form, after Close() has been called
+	Code() string
+}
+
+func (l *lexer) Next() Token {
+	token, ok := <-l.tokens
+	if !ok {
+		pos := ast.Position{
+			Offset: l.endOffset - 1,
+			Line:   l.endPos().line,
+			Column: l.endPos().column - 1,
+		}
+		token = Token{
+			Type: TokenEOF,
+			Range: ast.Range{
+				StartPos: pos,
+				EndPos:   pos,
+			},
+		}
+	}
+	return token
+}
+
+func (l *lexer) Close() {
+	l.cancelLexer()
+}
+
+func (l *lexer) Code() string {
+	return l.input
```

Comment:
From the documentation of `Code` it seemed to me at first that the function would only return the lexed code up-to when `Close` was called

Reply:
313b77b